packages/block-library/src/block/edit.js

```diff
@@ -105,16 +214,38 @@ export default function ReusableBlockEdit( {
 	} );
 
 	const innerBlocksProps = useInnerBlocksProps( blockProps, {
-		value: blocks,
 		layout,
-		onInput,
-		onChange,
-		renderAppender: blocks?.length
+		renderAppender: innerBlocks?.length
 			? undefined
 			: InnerBlocks.ButtonBlockAppender,
 	} );
 
+	// Sync the `dynamicContent` attribute from the updated blocks.
+	// `syncDerivedBlockAttributes` is an action that just like `updateBlockAttributes`
+	// but won't create an undo level.
+	// This can be abstracted into a `useSyncDerivedAttributes` hook if needed.
+	useEffect( () => {
+		const { getBlocks } = registry.select( blockEditorStore );
+		const { syncDerivedBlockAttributes } = unlock(
+			registry.dispatch( blockEditorStore )
+		);
+		let prevBlocks = getBlocks( patternClientId );
+		return registry.subscribe( () => {
+			const blocks = getBlocks( patternClientId );
+			if ( blocks !== prevBlocks ) {
+				prevBlocks = blocks;
+				syncDerivedBlockAttributes( patternClientId, {
+					dynamicContent: getDynamicContentFromBlocks(
+						blocks,
+						defaultValuesRef.current
+					),
+				} );
+			}
+		}, blockEditorStore );
+	}, [ patternClientId, registry ] );
```

Comment:
Any concerns over using `registry.subscribe`?

Feels like something where there's a lot of scope to explore different API ideas, like being able to subscribe to specific blocks, or proxying updates from one block to another.

Not something where we should hold back from shipping a first version though, but it'd be good to keep the conversation going.

Reply:
I think it's fine as it's what `useBlockSync` uses under the hood too. It would be good to have something designed for this type of use case, but most of the other solutions I tried break the undo stack or require more wiring to make them work. I think we can ask for some help from `@wordpress/data` and undo/redo history experts here. For now, I think this is an easy-to-understand enough implementation for our first version.